File: pi/userenv/src/hybrid/hello_hybrid.f90
program dummy

  use omp_lib
  use mpi

  implicit none

  character*MPI_MAX_PROCESSOR_NAME, save ::  pname
  integer mpierr, itd, gtd, nproc, nrank, grank, nlen

! Init MPI
    call MPI_Init(mpierr)

! Get Rank Size
    call MPI_COMM_Rank(MPI_COMM_WORLD, nrank, mpierr)
    call MPI_COMM_Size(MPI_COMM_WORLD, nproc, mpierr)

! Print rank
    call MPI_GET_PROCESSOR_NAME(pname, nlen, mpierr)

! Get Date hostname etc
    if (nrank==0) then
        call system('hostname && date && echo rank-pid $$')
    end if
    !if (nrank==0) then
       !call system('top -n 1 -b -H -u bthakur')
       !call system('pstree -ap -u bthakur')
    !end if
    call MPI_Barrier(MPI_COMM_WORLD, mpierr)

! OpenMP
    !$OMP PARALLEL PRIVATE(itd,gtd)
      itd= omp_get_thread_num()
      gtd= omp_get_num_threads()
      grank= nrank*gtd + itd
      write(*,'(4(a6,i6),a2,a8)')"Gid ", grank, " Im ", nrank, &
                                " of ", nproc, &
                                " thd", itd,    &
                                " on ", pname
    !$OMP FLUSH
    !$OMP BARRIER
      if (nrank==0 .and. itd==0) then
         call system('pstree -ap -u $USER')
         !call system('for f in `cat $PBS_NODEFILE|uniq`; do ssh $f pstree -ap -u bthakur; done;')
                
         call system('sleep 10')
      end if
    !$OMP FLUSH
    !$OMP END PARALLEL
    
    call MPI_Barrier(MPI_COMM_WORLD, mpierr)

! Finalize
    call MPI_Finalize(mpierr)
end

File: pi/userenv/src/hybrid/hello_hybrid.c
#include <stdio.h>
#include "mpi.h"
#include <omp.h>

int main(int argc, char *argv[]) {
    int numprocs, nrank, namelen;
    char processor_name[MPI_MAX_PROCESSOR_NAME];
    int gtd=0, itd=0, gid=0, np = 1;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &numprocs);
    MPI_Comm_rank(MPI_COMM_WORLD, &nrank);
    MPI_Get_processor_name(processor_name, &namelen);

#pragma omp parallel default(shared) private(itd, np)
    {
        gtd = omp_get_num_threads();
        itd = omp_get_thread_num();
        gid = nrank*gtd + itd;// global id
        printf("Gid %d from thread %d out of %d from process %d out of %d on %s\n",
                gid, itd, gtd, nrank, numprocs, processor_name);
        if (nrank==0 && itd==0) 
        {
            // system("pstree -ap -u $USER");
            //system("for f in `cat $PBS_NODEFILE|uniq`; do ssh $f pstree -ap -u $USER; done;");
            //system("sleep 10");
        }
    }

    if (nrank==0) 
        system("hostname && date && echo rank-pid $$");

    MPI_Finalize();
}
